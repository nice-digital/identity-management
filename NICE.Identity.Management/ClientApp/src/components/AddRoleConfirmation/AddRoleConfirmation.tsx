import React from "react";
import { Link } from "react-router-dom";

import { Alert } from "@nice-digital/nds-alert";

type AddRoleConfirmationProps = {
	id: number;
	fullName: string;
};

export const AddRoleConfirmation = (
	props: AddRoleConfirmationProps,
): React.ReactElement => {
	const { id, fullName } = props;

	return (
		<Alert type="success">
			<p>Your changes have been saved.</p>
			<Link to={`/users/${id}`}>Back to {fullName}</Link>
		</Alert>
	);
};
