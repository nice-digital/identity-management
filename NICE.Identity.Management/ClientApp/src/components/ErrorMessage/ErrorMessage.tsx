import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps): React.ReactElement => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there&apos;s been an error!</p>
			</>
		)}

		{process.env.NODE_ENV === "development" && (
			<dl>
				<dt>Message:</dt>
				<dd>{props.error.message}</dd>
				<dt>Stack:</dt>
				<dd>{props.error.stack}</dd>
			</dl>
		)}
	</Alert>
);
