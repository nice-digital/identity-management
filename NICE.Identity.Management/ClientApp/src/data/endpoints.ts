const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
