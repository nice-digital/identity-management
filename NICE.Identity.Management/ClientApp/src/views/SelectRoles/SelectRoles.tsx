import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";
import { type EnvironmentType, type ServiceType, type UserType, type RoleType, type WebsiteType } from "src/models/types";
import { fetchData } from "src/helpers/fetchData";
import { isDataError } from "src/helpers/isDataError";
import { Endpoints } from "src/data/endpoints";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { Button } from "@nice-digital/nds-button";

import { ErrorMessage } from "src/components/ErrorMessage/ErrorMessage";
import { AddRoleConfirmation } from "src/components/AddRoleConfirmation/AddRoleConfirmation";
import { ToggleCheckbox } from "src/components/ToggleCheckbox/ToggleCheckbox";

import "@nice-digital/nds-checkbox/scss/checkbox.scss";

type TParams = { id: string; serviceId: string; websiteId: string };

type SelectRolesProps = Record<string, unknown> & RouteComponentProps<TParams>;

type SelectRolesState = {
	user: UserType;
	roles: Array<RoleType>;
	service: ServiceType;
	website: WebsiteType;
	environment: EnvironmentType;
	hasBeenUpdated: boolean;
	error?: Error;
	isLoading: boolean;
	isButtonDisabled: boolean;
};

export class SelectRoles extends Component<SelectRolesProps, SelectRolesState> {
	constructor(props: SelectRolesProps) {
		super(props);
		this.state = {
			user: {} as UserType,
			roles: [],
			service: {} as ServiceType,
			website: {} as WebsiteType,
			environment: {} as EnvironmentType,
			hasBeenUpdated: false,
			isLoading: true,
			isButtonDisabled: false,
		};

		document.title = "NICE Accounts - Select role";
	}

	handleCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
		const checkbox = e.target;
		const roles = this.state.roles;

		roles.forEach((role) => {
			if (role.id.toString() === checkbox.value) {
				role.hasRole = !role.hasRole;
			}
		});

		this.setState({ roles });
	};

	handleFormSubmission = async (e: React.FormEvent): Promise<void> => {
		e.preventDefault();

		this.setState({ isButtonDisabled: true });

		const fetchOptions = {
			method: "PATCH",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ roles: this.state.roles }),
		};

		const roles = await fetchData(
			Endpoints.userRolesByWebsite(
				this.props.match.params.id,
				this.props.match.params.websiteId,
			),
			fetchOptions,
		);

		if (isDataError(roles)) {
			this.setState({ error: roles });
		} else {
			this.setState({ hasBeenUpdated: true });
		}

		this.setState({ isButtonDisabled: false });
	};

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const user = await fetchData(Endpoints.user(this.props.match.params.id));
		const userRolesByWebsite = await fetchData(
			Endpoints.userRolesByWebsite(
				this.props.match.params.id,
				this.props.match.params.websiteId,
			),
		);

		if (isDataError(user)) {
			this.setState({ error: user });
		}

		if (isDataError(userRolesByWebsite)) {
			this.setState({ error: userRolesByWebsite });
		}

		const { roles, service, website } = userRolesByWebsite;

		const environment = website ? website.environment : {};

		this.setState({
			user,
			roles,
			service,
			website,
			environment,
			isLoading: false,
		});
	}

	render(): JSX.Element {
		const {
			user,
			roles,
			service,
			environment,
			hasBeenUpdated,
			error,
			isLoading,
			isButtonDisabled,
		} = this.state;
		const { id, serviceId } = this.props.match.params;

		let nameBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			nameBreadcrumb = "Loading user details";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb
						data-qa-sel="breadcrumb-administration-link"
						to="/"
						elementType={Link}
					>
						Administration
					</Breadcrumb>
					<Breadcrumb
						data-qa-sel="breadcrumb-user-link"
						to="/users"
						elementType={Link}
					>
						Users
					</Breadcrumb>
					{error ? (
						<Breadcrumb>Error</Breadcrumb>
					) : (
						<Breadcrumb to={`/users/${id}`} elementType={Link}>
							{nameBreadcrumb}
						</Breadcrumb>
					)}
					<Breadcrumb to={`/users/${id}/services`} elementType={Link}>
						Select service
					</Breadcrumb>
					<Breadcrumb
						to={`/users/${id}/services/${serviceId}/environments`}
						elementType={Link}
					>
						Select environment
					</Breadcrumb>
					<Breadcrumb>Select role</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							preheading="Select roles for"
							heading={
								isLoading
									? "User details"
									: `${service.name} (${environment.name})`
							}
						/>

						{hasBeenUpdated && (
							<Grid>
								<GridItem cols={12} md={9}>
									<AddRoleConfirmation
										id={user.userId}
										fullName={`${user.firstName} ${user.lastName}`}
									/>
								</GridItem>
							</Grid>
						)}

						<Grid>
							<GridItem cols={12} md={9}>
								{isLoading ? (
									<p>Loading...</p>
								) : (
									<form onSubmit={this.handleFormSubmission}>
										{roles.map((role) => {
											const props = {
												id: role.id,
												name: role.name,
												description: role.description,
												isSelected: role.hasRole,
												isDisabled: isButtonDisabled,
												onCheckboxChange: this.handleCheckboxChange,
											};

											return <ToggleCheckbox {...props} key={role.id} />;
										})}
										<Button
											buttonType="submit"
											disabled={isButtonDisabled}
											className="mt--d"
										>
											{isButtonDisabled ? "Loading..." : "Save"}
										</Button>
									</form>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
