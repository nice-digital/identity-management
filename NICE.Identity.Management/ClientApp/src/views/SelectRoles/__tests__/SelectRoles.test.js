import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import singleUserRoles from "./singleUserRoles.json";
import { SelectRoles } from "../SelectRoles";

import { Endpoints } from "../../../data/endpoints";
import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";
import { AddRoleConfirmation } from "../../../components/AddRoleConfirmation/AddRoleConfirmation";

describe("SelectRoles", () => {
	const match = {
		params: { id: 1, serviceId: 1, websiteId: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = shallow(<SelectRoles match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = mount(<MemoryRouter><SelectRoles match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		await nextTick();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(2);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.user(match.params.id));
		expect(fetch.mock.calls[1][0]).toEqual(Endpoints.userRolesByWebsite(match.params.id, match.params.websiteId));
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = shallow(<SelectRoles match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when user fetchData function returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when user fetchData function returns 401 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 401 error", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should display confirmation message once fetchData patch is successfully complete", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").simulate("submit");
		await nextTick();
		wrapper.update();
		expect(wrapper.find(AddRoleConfirmation).exists()).toBe(true);
	});

	it("should disable button when form submitted", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		fetch.mockResponseOnce(JSON.stringify({}));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").simulate("submit");
		wrapper.update();
		expect(wrapper.find("button").props().disabled).toEqual(true);
		expect(wrapper.find("button").text()).toEqual("Loading...");
	});

	it("should disable all checkboxes when form submitted", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserRoles));
		fetch.mockResponseOnce(JSON.stringify({}));
		const wrapper = mount(
			<MemoryRouter>
				<SelectRoles match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").simulate("submit");
		wrapper.update();
		wrapper.find({ type: "checkbox" }).forEach(checkbox => {
			expect(checkbox.props().disabled).toEqual(true);
		});
	});
});
