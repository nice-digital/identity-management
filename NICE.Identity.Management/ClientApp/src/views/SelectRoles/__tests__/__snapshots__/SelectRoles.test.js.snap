// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Kristin Patrick
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services"
        >
          Select service
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services/1/environments"
        >
          Select environment
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select role
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Kristin Patrick"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/users/1/services",
        "name": "Select service"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/users/1/services/1/environments",
        "name": "Select environment"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Select role"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      <span
        class="page-header__pre-heading"
      >
        Select roles for
      </span>
      EPPI Reviewer (dev)
    </h1>
  </div>
  <div
    class="grid"
  >
    <div
      data-g="12 md:9"
    >
      <form>
        <div
          class="checkbox"
        >
          <input
            checked=""
            class="checkbox__input"
            id="Productcontributor"
            name="service-role"
            type="checkbox"
            value="991"
          />
          <label
            class="checkbox__label mb--0"
            for="Productcontributor"
          >
            Product contributor
          </label>
          <span
            class="checkbox__hint"
          >
            Description for Product contributor
          </span>
        </div>
        <div
          class="checkbox"
        >
          <input
            checked=""
            class="checkbox__input"
            id="Productmanager"
            name="service-role"
            type="checkbox"
            value="992"
          />
          <label
            class="checkbox__label mb--0"
            for="Productmanager"
          >
            Product manager
          </label>
          <span
            class="checkbox__hint"
          >
            Description for Product manager
          </span>
        </div>
        <div
          class="checkbox"
        >
          <input
            checked=""
            class="checkbox__input"
            id="Productadministrator"
            name="service-role"
            type="checkbox"
            value="993"
          />
          <label
            class="checkbox__label mb--0"
            for="Productadministrator"
          >
            Product administrator
          </label>
          <span
            class="checkbox__hint"
          >
            Description for Product administrator
          </span>
        </div>
        <div
          class="checkbox"
        >
          <input
            class="checkbox__input"
            id="Producteditor"
            name="service-role"
            type="checkbox"
            value="994"
          />
          <label
            class="checkbox__label mb--0"
            for="Producteditor"
          >
            Product editor
          </label>
          <span
            class="checkbox__hint"
          >
            Description for Product editor
          </span>
        </div>
        <button
          class="btn mt--d"
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`should show error message when user fetchData function returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services"
        >
          Select service
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services/1/environments"
        >
          Select environment
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select role
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/users/1/services",
        "name": "Select service"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/users/1/services/1/environments",
        "name": "Select environment"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Select role"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when user fetchData function returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services"
        >
          Select service
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services/1/environments"
        >
          Select environment
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select role
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/users/1/services",
        "name": "Select service"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/users/1/services/1/environments",
        "name": "Select environment"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Select role"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when user roles fetchData function returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services"
        >
          Select service
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services/1/environments"
        >
          Select environment
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select role
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/users/1/services",
        "name": "Select service"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/users/1/services/1/environments",
        "name": "Select environment"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Select role"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when user roles fetchData function returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services"
        >
          Select service
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1/services/1/environments"
        >
          Select environment
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Select role
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/users/1/services",
        "name": "Select service"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/users/1/services/1/environments",
        "name": "Select environment"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Select role"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;
